### PR TITLE
Add CloudStack into eksa controller manager

### PR DESCRIPTION
**Pull request by wanyufe**
*Issue #, if available:*

*Description of changes:*
This is part of PRs that merge from embargo to public repo. This PR is about to add CloudStack into eks-a controller.
*Testing (if applicable):*
Unit testing
Manual test like creating/upgrading Cloudstack cluster are still on-going.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/controllers/resource/testdata/cloudstackMachineDeployment.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+apiVersion: cluster.x-k8s.io/v1alpha3
```
Comment (vignesh-goutham) [inline controllers/controllers/resource/testdata/cloudstackMachineDeployment.yaml:1]: all these should be v1beta1. Please change all the v1alpha3s to v1beta1

File: controllers/controllers/resource/testdata/eksa-cluster-cloudstack.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+kind: CloudStackMachineConfig
+metadata:
+  creationTimestamp: null
+  name: test_cluster
+spec:
+  computeoffering: large
+  diskoffering: small
+  osFamily: ubuntu
+  template: rhel8-1.20
+  details:
```
Comment (vignesh-goutham) [inline controllers/controllers/resource/testdata/eksa-cluster-cloudstack.yaml:41]: we changed this to userCustomDetails, please verify the correct api in the codebase now.

File: controllers/controllers/resource/testdata/eksa-cluster-vsphere.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
```
Comment (vignesh-goutham) [inline controllers/controllers/resource/testdata/eksa-cluster-vsphere.yaml:1]: We shouldnt be adding vpshere files. It looks like this PR needs a rebase on top of main on this repo.

File: pkg/api/v1alpha1/testdata/not_parseable_cluster_cloudstack.yaml [hunk 4]
```diff
@@ -37,12 +37,9 @@ spec:
...
   domain: "domain1"
   zones:
-    - zone:
-        value: "zone1"
-        type: "Name"
+    - name: "zone1"
       network:
-        value: "net1"
-        type: "Name"
+        name: "net1"
```
Comment (vignesh-goutham) [inline pkg/api/v1alpha1/testdata/not_parseable_cluster_cloudstack.yaml:42]: We should add a place where we test out `id` as opposed to `name`

File: pkg/providers/cloudstack/testdata/expected_results_main_md.yaml [hunk 5]
```diff
@@ -65,9 +65,11 @@ metadata:
 spec:
   template:
     spec:
-      offering: m4-large
-      template: centos7-k8s-118
-      details:
-        foo: bar
+      offering:
+        id:
```
Comment (vignesh-goutham) [inline pkg/providers/cloudstack/testdata/expected_results_main_md.yaml:69]: Lets remove the id so that its not confusing.

**Review by vignesh-goutham — COMMENTED** [review]
It looks like the PR needs to be rebased on top of main to get in the latest changes.

File: pkg/providers/cloudstack/cloudstack.go [hunk 6]
```diff
@@ -78,16 +79,83 @@ func (p *cloudstackProvider) UpdateSecrets(ctx context.Context, cluster *types.C
...
+
+	if nSpec.Domain != oSpec.Domain {
+		return fmt.Errorf("spec.domain is immutable. Previous value %s, new value %s", oSpec.Domain, nSpec.Domain)
+	}
+	if nSpec.Domain != oSpec.Domain {
+		return fmt.Errorf("spec.domain is immutable. Previous value %s, new value %s", oSpec.Account, nSpec.Account)
+	}
+	//if !reflect.DeepEqual(nSpec.Zones, oSpec.Zones) {
+	//	return fmt.Errorf("spec.zones are immutable. Previous value %s, new value %s", oSpec.Zones, nSpec.Zones)
+	//}
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack.go:134]: I was getting false positives with this validation. Could you try re-enabling it and seeing if it works for you?

File: pkg/providers/cloudstack/cloudstack.go [hunk 7]
```diff
@@ -351,7 +409,34 @@ func (p *cloudstackProvider) SetupAndValidateCreateCluster(ctx context.Context,
...
+	}
+
+	if err := p.setupSSHAuthKeysForUpgrade(); err != nil {
+		return fmt.Errorf("failed setup and validations: %v", err)
+	}
+	err = p.setupSSHAuthKeysForUpgrade()
+	if err != nil {
+		return fmt.Errorf("failed setup and validations: %v", err)
+	}
+	err = p.validateMachineConfigsNameUniqueness(ctx, cluster, clusterSpec)
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack.go:435]: should this go in the validator?

File: pkg/providers/cloudstack/cloudstack.go [hunk 8]
```diff
@@ -476,10 +648,6 @@ func buildTemplateMapCP(clusterSpec *cluster.Spec, datacenterConfigSpec v1alpha1
 
 		// Add no-proxy defaults
 		noProxyList = append(noProxyList, common.NoProxyDefaults...)
-		cloudStackManagementApiEndpointHostname, err := getHostnameFromUrl(managementApiEndpoint)
-		if err == nil {
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack.go:480]: after further discussion, we agreed to leave the no-proxy details in, and grab the managementApiEndpoint from the cluster spec, right?

File: pkg/providers/cloudstack/config/template-cp.yaml [hunk 9]
```diff
@@ -304,14 +353,12 @@ metadata:
...
   template:
     spec:
-      offering: {{.cloudstackEtcdComputeOffering}}
-      template: {{.cloudstackEtcdTemplateOffering}}
-      {{- if .cloudstackEtcdCustomDetails }}
-      details:
-      {{- range $key, $value := .cloudstackEtcdCustomDetails}}
-        {{$key}}: {{$value}}
-      {{- end}}
-      {{- end}}
```
Comment (maxdrib) [inline pkg/providers/cloudstack/config/template-cp.yaml:314]: leave custom details in

**Review by maxdrib — CHANGES_REQUESTED** [review]

File: pkg/providers/cloudstack/testdata/expected_results_main_md.yaml (hunk 5 above)
Comment (wanyufe) [reply pkg/providers/cloudstack/testdata/expected_results_main_md.yaml:69]: yes, we have leave it in in our current template-md implementation. To remove it, we need to put conditional statement in template-cd/md.yaml files. 

This file is a intermediate file and customer could not see them. In the customer facing file, the id/name will not (does not) show up if they are empty.

File: controllers/controllers/clusters/cloudstack.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+	"github.com/aws/eks-anywhere/controllers/controllers/reconciler"
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+)
+
+type CloudStackReconciler struct {
+	*providerClusterReconciler
+}
+
+func NewCloudStackReconciler() *CloudStackReconciler {
```
[the remark this answers is not in the document]
Comment (wanyufe) [reply controllers/controllers/clusters/cloudstack.go:14]: It is not used anywhere now. This is designed to be used in the future version of cluster_controller.go

Currently, cloudstack is using cluster_controller_legacy.go

This whole file is not used yet, I am thinking I can remove it.

File: controllers/controllers/resource/fetcher.go [hunk 11]
```diff
@@ -68,6 +73,7 @@ func (r *CapiResourceFetcher) FetchObjectByName(ctx context.Context, name string
 }
 
 func (r *CapiResourceFetcher) FetchObject(ctx context.Context, objectKey types.NamespacedName, obj client.Object) error {
+	r.log.Info("FetchObject", "key", objectKey, "obj", obj)
```
[the remark this answers is not in the document]
Comment (wanyufe) [reply controllers/controllers/resource/fetcher.go:76]: agree

[the remark this answers is not in the document]
Comment (wanyufe) [reply controllers/controllers/resource/fetcher.go:76]: agree

File: controllers/controllers/resource/reconciler_test.go [hunk 12]
```diff
@@ -514,13 +544,448 @@ func TestClusterReconcilerReconcile(t *testing.T) {
...
+				fetcher.EXPECT().FetchObject(gomock.Any(), gomock.Any(), gomock.Any()).Do(func(ctx context.Context, objectKey types.NamespacedName, obj client.Object) {
+					cluster := obj.(*anywherev1.CloudStackDatacenterConfig)
+					cluster.SetName(objectKey.Name)
+					cluster.SetNamespace(objectKey.Name)
+					cluster.Spec = datacenterSpec.Spec
+					assert.Equal(t, objectKey.Name, "test_cluster", "expected Name to be test_cluster")
+				}).Return(nil)
+
+				existingVSDatacenter := &anywherev1.CloudStackDatacenterConfig{}
+				existingVSDatacenter.Spec = datacenterSpec.Spec
```
Comment (maxdrib) [inline controllers/controllers/resource/reconciler_test.go:727]: csdatacenter?

File: pkg/providers/cloudstack/config/secret.yaml [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+metadata:
+  name: eksa-system
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: eksa-license
+  namespace: eksa-system
+stringData:
+  license: "{{.eksaLicense}}"
```
Comment (maxdrib) [inline pkg/providers/cloudstack/config/secret.yaml:12]: is this file used anywhere? can we remove it?

**Review by maxdrib — COMMENTED** [review]
Minor comments. Were we planning to add apiManagementEndpoint in this PR as well?

File: pkg/providers/cloudstack/config/secret.yaml (hunk 13 above)
Comment (vignesh-goutham) [reply pkg/providers/cloudstack/config/secret.yaml:12]: We dont use this anywhere functional yet, but this is used when the support bundle is run.

**Review by vignesh-goutham — COMMENTED** [review]
Few comments. Overall looks good.

File: controllers/controllers/resource/fetcher.go [hunk 14]
```diff
@@ -422,6 +511,43 @@ func MapMachineTemplateToVSphereMachineConfigSpec(vsMachineTemplate *vspherev1.V
 	return vsSpec, nil
 }
 
+func MapClusterToCloudStackDatacenterConfigSpec(csCluster *cloudstackv1.CloudStackCluster) (*anywherev1.CloudStackDatacenterConfig, error) {
```
Comment (jiayiwang7) [inline controllers/controllers/resource/fetcher.go:514]: I dont think `error` is needed in return?

File: controllers/controllers/resource/fetcher.go [hunk 15]
```diff
@@ -412,6 +501,43 @@ func MapMachineTemplateToVSphereMachineConfigSpec(vsMachineTemplate *vspherev1.V
...
+		Name: csMachineTemplate.Spec.Spec.Spec.Offering.Name,
+	}
+	csSpec.Spec.Template = anywherev1.CloudStackResourceIdentifier{
+		Id:   csMachineTemplate.Spec.Spec.Spec.Template.ID,
+		Name: csMachineTemplate.Spec.Spec.Spec.Template.Name,
+	}
+
+	csSpec.Spec.AffinityGroupIds = csMachineTemplate.Spec.Spec.Spec.AffinityGroupIDs
+
+	// TODO: Details, Users (these fields are immutable)
```
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/resource/fetcher.go:547]: immutability validation is in webhook, not in main controller code

File: controllers/controllers/resource/reconciler.go [hunk 16]
```diff
@@ -96,6 +102,41 @@ func (cor *clusterReconciler) Reconcile(ctx context.Context, objectKey types.Nam
...
+		workerCsmcs := make(map[string]anywherev1.CloudStackMachineConfig, len(cs.Spec.WorkerNodeGroupConfigurations))
+		err := cor.FetchObject(ctx, types.NamespacedName{Namespace: objectKey.Namespace, Name: cs.Spec.DatacenterRef.Name}, csdc)
+		if err != nil {
+			return err
+		}
+		err = cor.FetchObject(ctx, types.NamespacedName{Namespace: objectKey.Namespace, Name: cs.Spec.ControlPlaneConfiguration.MachineGroupRef.Name}, cpCsmc)
+		if err != nil {
+			return err
+		}
+		if len(cs.Spec.WorkerNodeGroupConfigurations) != 1 {
```
Comment (jiayiwang7) [inline controllers/controllers/resource/reconciler.go:119]: so no multiple worker node groups support?

File: controllers/controllers/resource/reconciler.go [hunk 17]
```diff
@@ -96,6 +102,41 @@ func (cor *clusterReconciler) Reconcile(ctx context.Context, objectKey types.Nam
...
+			return fmt.Errorf("expects WorkerNodeGroupConfigurations's length to be 1, but found %d", len(cs.Spec.WorkerNodeGroupConfigurations))
+		}
+		for _, workerNodeGroupConfiguration := range cs.Spec.WorkerNodeGroupConfigurations[:1] {
```
Comment (jiayiwang7) [inline controllers/controllers/resource/reconciler.go:122]: since the worker node group length has to be 1, why do u still need this loop?

File: controllers/controllers/resource/template.go [hunk 18]
```diff
@@ -159,7 +166,114 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
...
 
-	return generateTemplateResources(templateBuilder, clusterSpec, workloadTemplateNames, kubeadmconfigTemplateNames, cpOpt)
+	return generateTemplateResources(templateBuilder, clusterSpec, workloadTemplateNames, kubeadmconfigTemplateNames, cpOpt, nil)
+}
+
+func (r *CloudStackTemplate) TemplateResources(ctx context.Context, eksaCluster *anywherev1.Cluster, clusterSpec *cluster.Spec, csdc anywherev1.CloudStackDatacenterConfig, cpCsmc, etcdCsmc anywherev1.CloudStackMachineConfig, workerCsmcs map[string]anywherev1.CloudStackMachineConfig) ([]*unstructured.Unstructured, error) {
+	if len(workerCsmcs) > 1 {
+		return nil, fmt.Errorf("multiple worker node group configuration is not supported in CloudStack provider")
+	}
+	workerNodeGroupMachineSpecs := make(map[string]anywherev1.CloudStackMachineConfigSpec, len(workerCsmcs))
```
Comment (jiayiwang7) [inline controllers/controllers/resource/template.go:176]: again why is this needed if not supporting multi worker node groups

File: controllers/controllers/resource/template.go [hunk 19]
```diff
@@ -159,7 +166,114 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
 		values["eksaVspherePassword"] = string(passwordBytes)
 	}
 
-	return generateTemplateResources(templateBuilder, clusterSpec, workloadTemplateNames, kubeadmconfigTemplateNames, cpOpt)
+	return generateTemplateResources(templateBuilder, clusterSpec, workloadTemplateNames, kubeadmconfigTemplateNames, cpOpt, nil)
+}
+
+func (r *CloudStackTemplate) TemplateResources(ctx context.Context, eksaCluster *anywherev1.Cluster, clusterSpec *cluster.Spec, csdc anywherev1.CloudStackDatacenterConfig, cpCsmc, etcdCsmc anywherev1.CloudStackMachineConfig, workerCsmcs map[string]anywherev1.CloudStackMachineConfig) ([]*unstructured.Unstructured, error) {
+	if len(workerCsmcs) > 1 {
```
Comment (jiayiwang7) [inline controllers/controllers/resource/template.go:173]: seems all below code handles multiple worker node groups. why still have this `len(workerCsmcs) > 1` check?

File: controllers/controllers/resource/template.go (hunk 18 above)
Comment (jiayiwang7) [inline controllers/controllers/resource/template.go:176]: i understand most of the code is copied from vsphere. but it really becomes hard to follow what this method is doing. is it possible to break it down by adding smaller methods like
- workerNodeGroupMachineSpec()
- controlPlaneTemplateName()
- workerTemplateName()
- etcdTemplateName()
- ...

File: controllers/controllers/resource/template.go [hunk 20]
```diff
@@ -159,7 +166,114 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
...
+	} else {
+		cp, err := r.ControlPlane(ctx, eksaCluster)
+		if err != nil {
+			return nil, err
+		}
+		controlPlaneTemplateName = cp.Spec.MachineTemplate.InfrastructureRef.Name
+	}
+
+	kubeadmconfigTemplateNames := make(map[string]string, len(clusterSpec.Spec.WorkerNodeGroupConfigurations))
+	workloadTemplateNames := make(map[string]string, len(clusterSpec.Spec.WorkerNodeGroupConfigurations))
```
Comment (jiayiwang7) [inline controllers/controllers/resource/template.go:206]: nit: rename to `workersTemplateNames`

File: controllers/controllers/resource/template.go [hunk 21]
```diff
@@ -159,7 +166,114 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
...
+
+	cpOpt := func(values map[string]interface{}) {
+		values["controlPlaneTemplateName"] = controlPlaneTemplateName
+		values["cloudstackControlPlaneSshAuthorizedKey"] = sshAuthorizedKey(cpCsmc.Spec.Users)
+		values["cloudstackEtcdSshAuthorizedKey"] = sshAuthorizedKey(etcdCsmc.Spec.Users)
+		values["etcdTemplateName"] = etcdTemplateName
+	}
+
+	workersOpt := func(values map[string]interface{}) {
+		values["workloadTemplateName"] = workloadTemplateNames[clusterSpec.Spec.WorkerNodeGroupConfigurations[0].Name]
```
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/resource/template.go:272]: It's quite messy that part of the cloudstack code supports multi worker node group while others not.. like here workersOpts should not be needed in other providers since it's passed in `workloadTemplateNames`. You are introducing redundant args for this "partial" support.

File: controllers/controllers/resource/template.go [hunk 22]
```diff
@@ -159,7 +166,114 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
...
+		values["cloudStackWorkerSshAuthorizedKey"] = sshAuthorizedKey(workerNodeGroupMachineSpecs[clusterSpec.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name].Users)
```
Comment (jiayiwang7) [inline controllers/controllers/resource/template.go:273]: Is `cloudStackWorkerSshAuthorizedKey` required? I remember we removed this from vsphere controller recently. Can u pls check the vsphere code and see if this is still needed

File: pkg/clustermanager/cluster_manager.go [hunk 23]
```diff
@@ -510,6 +519,52 @@ func (c *ClusterManager) EKSAClusterSpecChanged(ctx context.Context, cluster *ty
 				return true, nil
 			}
 		}
+	case v1alpha1.CloudStackDatacenterKind:
```
Comment (jiayiwang7) [inline pkg/clustermanager/cluster_manager.go:522]: `EKSAClusterSpecChanged` should be a provider method to avoid switch statement

**Comment by vivek-koppuru** [comment]
Are we putting this behind a feature flag somehow? This will build controller images for all users vs for the CLI where I think it is gated for the CLI right?

File: config/crd/bases/anywhere.eks.amazonaws.com_cloudstackmachineconfigs.yaml [hunk 24]
```diff
@@ -56,26 +56,20 @@ spec:
                   size including number of CPU’s, memory, and CPU speed. It can either
                   be specified as a UUID or name
                 properties:
-                  type:
+                  id:
```
Comment (vivek-koppuru) [inline config/crd/bases/anywhere.eks.amazonaws.com_cloudstackmachineconfigs.yaml:59]: Is this the PR where we are refactoring the changes from `type`/`value` to `id`/`name`? If so, can we split the PR into logical chunks so that it is easier to review?

File: controllers/controllers/resource/fetcher.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (vivek-koppuru) [reply controllers/controllers/resource/fetcher.go:547]: I think the todo was added more to support changing that in the future too, not that we are checking for immutability.

File: controllers/controllers/resource/reconciler.go (hunk 16 above)
Comment (vivek-koppuru) [reply controllers/controllers/resource/reconciler.go:119]: I think this was talked about in the past, but I don't remember if the plan was to release without this support. We should keep cloudstack behind a feature flag if we don't support multiple worker node groups imo

File: controllers/controllers/resource/template.go (hunk 22 above)
Comment (vivek-koppuru) [reply controllers/controllers/resource/template.go:273]: Yea we get it from the creds secret now: https://github.com/aws/eks-anywhere/blob/main/controllers/controllers/resource/template.go#L140

File: docs/content/en/docs/reference/changelog.md [hunk 25]
```diff
@@ -30,7 +30,7 @@ menu:
 - Cilium CNI will now be upgraded by the `upgrade cluster` command [#326](https://github.com/aws/eks-anywhere/issues/326)
 
 ### Changed
-- EKS-A now uses Cluster API (CAPI) v1.0.1 and v1beta1 manifests, upgrading from v0.3.23 and v1alpha3 manifests.
+- EKS-A now uses Cluster API (CAPI) v1.0.1 and v1beta1 manifests, upgrading from v0.3.23 and v1beta1 manifests.
```
Comment (vivek-koppuru) [inline docs/content/en/docs/reference/changelog.md:33]: Why is this being changed?

File: pkg/clustermanager/cluster_manager.go [hunk 26]
```diff
@@ -371,6 +374,12 @@ func (c *ClusterManager) UpgradeCluster(ctx context.Context, managementCluster,
 		return fmt.Errorf("error running post control plane upgrade operations: %v", err)
 	}
 
+	logger.V(3).Info("Waiting for control plane upgrade to be in progress")
+	err = c.clusterClient.WaitForControlPlaneNotReady(ctx, managementCluster, ctrlPlaneInProgressStr, newClusterSpec.Name)
```
Comment (vivek-koppuru) [inline pkg/clustermanager/cluster_manager.go:378]: This would affect other providers too right? What is the reason for adding this?

File: pkg/executables/cmk.go [hunk 27]
```diff
@@ -265,22 +264,22 @@ func (c *Cmk) ValidateNetworkPresent(ctx context.Context, domainId string, zoneR
 	networks := response.CmkNetworks
 
 	// filter by network name -- cmk does not support name= filter
-	if zoneRef.Network.Type == v1alpha1.Name {
+	if len(zone.Network.Name) > 0 {
```
[the remark this answers is not in the document]
Comment (vivek-koppuru) [reply pkg/executables/cmk.go:267]: We should make sure we validate that both aren't empty as well. If we do, I guess this is okay but I would prefer having an if else than two ifs here

File: pkg/providers/cloudstack/config/secret.yaml (hunk 13 above)
Comment (vivek-koppuru) [reply pkg/providers/cloudstack/config/secret.yaml:12]: It's stored as vsphere only because this is not a file we need for docker as that is for experimental purposes, but if we want to go with a license approach for cloudstack, we would apply it here like we do for vsphere. Probably need to generalize it more, but I am assuming this can be a follow up action item to address based on if we even have a license to give people for this provider

https://github.com/aws/eks-anywhere/blob/main/pkg/providers/vsphere/vsphere.go#L1028

File: pkg/providers/cloudstack/testdata/expected_results_main_md.yaml (hunk 5 above)
Comment (vivek-koppuru) [reply pkg/providers/cloudstack/testdata/expected_results_main_md.yaml:69]: I think we leave it if we don't make the template conditional. There are any edge cases we need to worry about by doing this right? Otherwise it might be worth making it conditional in the template.

File: pkg/providers/provider.go [hunk 28]
```diff
@@ -50,7 +50,7 @@ type BuildMapOption func(map[string]interface{})
 
 type TemplateBuilder interface {
 	GenerateCAPISpecControlPlane(clusterSpec *cluster.Spec, buildOptions ...BuildMapOption) (content []byte, err error)
-	GenerateCAPISpecWorkers(clusterSpec *cluster.Spec, workloadTemplateNames, kubeadmconfigTemplateNames map[string]string) (content []byte, err error)
+	GenerateCAPISpecWorkers(clusterSpec *cluster.Spec, workloadTemplateNames, kubeadmconfigTemplateNames map[string]string, buildOptions ...BuildMapOption) (content []byte, err error)
```
Comment (vivek-koppuru) [inline pkg/providers/provider.go:53]: For changes like these which affect other providers too, I would prefer separate PRs for high visibility among the team on what is being proposed and why.

File: controllers/controllers/resource/template.go (hunk 22 above)
Comment (jiayiwang7) [reply controllers/controllers/resource/template.go]: i see, i think workerOpts is inevitable since cloudstack is supporting worker nodes SSH upgrade through controller, while vsphere we dont support it yet. @vivek-koppuru

Comment (vivek-koppuru) [reply controllers/controllers/resource/template.go]: Oh right that's fair, is this something that was requested?

**Comment by vivek-koppuru** [comment]
Are we changing the branch of this PR now? Or opening a different PR?

File: controllers/controllers/resource/template.go [hunk 29]
```diff
@@ -159,7 +166,138 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
...
+	oldCsdc, err := r.ExistingCloudStackDatacenterConfig(ctx, eksaCluster, clusterSpec.Spec.WorkerNodeGroupConfigurations[0])
+	if err != nil {
+		return nil, err
+	}
+	oldCpCsmc, err := r.ExistingCloudStackControlPlaneMachineConfig(ctx, eksaCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	var controlPlaneTemplateName string
```
Comment (jiayiwang7) [inline controllers/controllers/resource/template.go:193]: `func getControlPlaneTemplateName()`?

File: controllers/controllers/resource/template.go [hunk 30]
```diff
@@ -159,7 +166,114 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
...
+		}
+
+		csmc := workerCsmcs[workerNodeGroupConfiguration.MachineGroupRef.Name]
+		oldCsmc, err := r.ExistingCloudStackWorkerMachineConfig(ctx, eksaCluster, workerNodeGroupConfiguration)
+		if err != nil {
+			return nil, err
+		}
+		updateWorkloadTemplate := cloudstack.AnyImmutableFieldChanged(oldCsdc, &csdc, oldCsmc, &csmc)
+		if updateWorkloadTemplate {
+			workloadTemplateName := common.WorkerMachineTemplateName(clusterName, workerNodeGroupConfiguration.Name, r.now)
```
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/resource/template.go:289]: workerTemplateName?

File: pkg/clustermanager/cluster_manager.go (hunk 23 above)
Comment (jiayiwang7) [reply pkg/clustermanager/cluster_manager.go:522]: can u pls create a tt to follow up on this?

**Comment by wanyufe** [comment]
duplicate with another PR https://github.com/aws/eks-anywhere/pull/1675
